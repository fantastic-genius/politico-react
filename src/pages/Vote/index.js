import React, { useEffect, useState } from 'react';
import { connect } from  'react-redux';
import DashboardContainer from '@components/DashboardContainer';
import { getVotedCandidates, getOfficeCandidates, voteCandidate, cleanUp } from '@actions/Vote';
import getOffices from '@actions/Office';
import CandidateRow from '@components/CandidateRow';
import { callToast } from '@components/Toast'

const Vote = (props) => {

  const {
    voted,
    offices,
    votedCandidates,
    candidates,
    candidatesLoading,
    candidatesLoaded,
    error,
    token,
    userId
  } = props;
  const [officeId, setOfficeId] = useState(null);
  const loadOffices = offices ? 
      offices.map(office => {
        if(votedCandidates && votedCandidates.length > 0){
            let isOfficeVoted = false
            votedCandidates.forEach(votedCandidate => {
                if(votedCandidate.officeid === office.id){
                  isOfficeVoted = true
                }
            })

            if(isOfficeVoted !== true){
                return (<option value={office.id} key={office.id}>{office.name}</option>)
            }
        }else{
            return (<option value={office.id} key={office.id}>{office.name}</option>);
        }
    }) : '';

  const onVote = (e) => {
    e.preventDefault();
    const button = e.target;
    const candidateId = button.dataset.candidateid;
    const values = {
      office: officeId,
      candidate: candidateId
    }
    props.voteCandidate(token, values);
  }

  const loadCandidates = candidatesLoaded && candidates.length ?
    candidates.map(candidate => (
      <CandidateRow key={candidate.id} candidate={candidate} onVote={onVote} />
      )) : (<tr></tr>);
  
  const onChange = (e) => {
    e.persist();
    const officId = e.target.value;
    setOfficeId(officId);
    /* istanbul ignore next */
    props.getOfficeCandidates(token, officId)
  }
  useEffect(() => {
    props.cleanUp();
    if(userId){
      /* istanbul ignore next */
      props.getVotedCandidates(userId, token)
      props.getOffices(token);
    }

    if (voted) {
      /* istanbul ignore next */
      callToast('success', 'You have successfully voted');
    }else if(error){
      /* istanbul ignore next */
      callToast('error', error);
    }
  },[userId, voted])

  return (
    <DashboardContainer>
      <article>
        <div className="title-con">
          <h3>Vote Choice Candidate</h3>
          <hr />
        </div>
        <div id="alert"></div>
        <div className="select-con">
          <label htmlFor="office">Political Office</label>
          <select name="office" id="office-select" onChange={onChange}>
            <option>Select an Office</option>
            {loadOffices}
          </select>
        </div>
        <div>
          {candidatesLoading ? <h3>loading...</h3> :
          <table id="candidates-table">
            <thead>
              <tr>
                <td></td>
                <td>Photo</td>
                <td>Politician Name</td>
                <td>Party</td>
                <td>Party Logo</td>
                <td>Action</td>
              </tr>
            </thead>
            <tbody>
              {loadCandidates}
            </tbody>
          </table>
          }
          {!candidates.length && !candidatesLoading ? (<h4>No Candidate for this office</h4>) : ''}
        </div>    
      </article>
    </DashboardContainer>
  )
}

const mapStateToProps = state => ({
  voted: state.vote.voted,
  candidatesLoading: state.vote.candidatesLoading,
  candidatesLoaded: state.vote.candidatesLoaded,
  votedCandidates: state.vote.votedCandidates,
  candidates: state.vote.candidates,
  error: state.vote.error,
  token: state.auth.token,
  userId: state.auth.user.id,
  offices: state.office.offices
});

export default connect(mapStateToProps, {
  getVotedCandidates, getOfficeCandidates, voteCandidate, getOffices, cleanUp
})(Vote);
