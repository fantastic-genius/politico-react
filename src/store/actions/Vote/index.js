import {axiosCall} from '@src/utils';
import * as types from '@actions/actionTypes';

export const getVotedOffices = (userId, token) => async (dispatch) => {
  dispatch({type: types.GET_VOTED_OFFICES_PENDING});

  try {
    const { data } = await axiosCall({
      path: `/api/v1/votes/${userId}/user`, method: 'get', token
    });
    dispatch({type: types.GET_VOTED_OFFICES_FULFILLED, payload: data});
  } catch (err) {
    const { response } = err;
    const error = response.data || response;
    dispatch({type: types.GET_VOTED_OFFICES_REJECTED, payload: error});
  }
};

export const getOfficeCandidates = (token, officeId) => async (dispatch) => {
  dispatch({type: types.GET_OFFICE_CANDIDATES_PENDING});

  try {
    const { data } = await axiosCall({
      path: `/api/v1/offices/${officeId}/candidates`, method: 'get', token
    });

    dispatch({type: types.GET_OFFICE_CANDIDATES_FULFILLED, payload: data});
  } catch (err) {
    const { response } = err;
    const error = response.data || response;
    dispatch({type: types.GET_OFFICE_CANDIDATES_REJECTED, payload: error});
  }
};

export const voteCandidate = (token, values) => async (dispatch) => {
  dispatch({type: types.VOTE_CANDIDATE_PENDING});

  try {
    const { data } = await axiosCall({
      path: '/api/v1/votes', method: 'post', payload: values, token
    });

    dispatch({type: types.VOTE_CANDIDATE_FULFILLED, payload: data});
  } catch (err) {
    const { response } = err;
    const error = response.data || response;
    dispatch({type: types.VOTE_CANDIDATE_REJECTED, payload: error});
  }
};

export const cleanUp = () => ({type: types.VOTE_CLEAN_UP});

