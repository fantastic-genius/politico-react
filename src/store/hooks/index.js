import { useEffect } from 'react';
import  { setUser } from '@actions/Auth';

const useSetUser = ({ dispatch }) => {
  useEffect(() => {
    dispatch(setUser());
  }, []);
};

export default useSetUser;