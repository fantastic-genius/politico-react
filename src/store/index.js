import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { createLogger as logger } from 'redux-logger';
import promise from 'redux-promise-middleware';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import reducer from './reducers';

const middleware = applyMiddleware(promise, thunk, logger());
const store = createStore(reducer, composeWithDevTools(middleware));

export default store;
