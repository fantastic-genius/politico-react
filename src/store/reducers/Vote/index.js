import * as types from '@actions/actionTypes';

const initialState = {
  votedCandidatesLoading: false,
  votedCandidatesLoaded: false,
  candidatesLoading: false,
  candidatesLoaded: false,
  voting: false,
  voted: false,
  votedCandidates: [],
  candidates: [],
  error: null
}

const voteReducer = (state=initialState, {type, payload}) => {
  switch (type) {
    case types.GET_VOTED_CANDIDATES_PENDING:
      return {
        ...state,
        votedCandidatesLoading: true,
      };
    case types.GET_VOTED_CANDIDATES_FULFILLED:
      return {
        ...state,
        votedCandidatesLoading: false,
        votedCandidatesLoaded: true,
        votedCandidates: payload
      };
    case types.GET_VOTED_CANDIDATES_REJECTED:
      return {
        ...state,
        votedCandidatesLoading: false,
        error: payload,
      };
    case types.GET_OFFICE_CANDIDATES_PENDING:
      return {
        ...state,
        candidatesLoading: true
      };
    case types.GET_OFFICE_CANDIDATES_FULFILLED:
      return {
        ...state,
        candidatesLoading: false,
        candidatesLoaded: true,
        candidates: payload
      };
    case types.GET_OFFICE_CANDIDATES_REJECTED:
      return {
        ...state,
        candidatesLoading: false,
        error: payload,
        candidates: [],
      };
    case types.VOTE_CANDIDATE_PENDING:
      return {
        ...state,
        voting: true,
      };
    case types.VOTE_CANDIDATE_FULFILLED:
      return {
        ...state,
        voting: false,
        voted: true,
      };
    case types.VOTE_CANDIDATE_REJECTED:
      return {
        ...state,
        voting: false,
        error: payload
      }
    case types.VOTE_CLEAN_UP:
      return {
        ...initialState
      }
    default:
      return state;
  }
}

export default voteReducer;
