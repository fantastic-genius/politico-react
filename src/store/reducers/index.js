import { combineReducers } from 'redux';
import auth from '@reducers/Auth';
import vote from '@reducers/Vote';
import office from '@reducers/Office';

const reducers = combineReducers({
    auth,
    vote,
    office,
});

export default reducers;
