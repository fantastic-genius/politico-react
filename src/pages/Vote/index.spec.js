import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux';
import store from '@src/store';
import Vote from './index';

describe('<Vote />', () => {
  describe('Rendering', () => {
    it('should render Vote page correctly', () => {
      const component = shallow(
        <Router>
          <Vote/>
        </Router>
      );
      expect(component).toMatchSnapshot();
    });

    it('should render Vote page correctly with its children when candidates is loading', () => {
        const component = mount(
          <Provider store={store}>
            <Router>
              <Vote candidatesLoading />
            </Router>
          </Provider>
        );
        expect(component).toMatchSnapshot();
        expect(component.find('select')).toHaveLength(1);
    });

    it('should render Vote page correctly with its children when candidates is loaded', () => {
      const candidates = [{
        id: 1,
        firstname: 'john',
        lastname: 'mark',
        passportUrl: 'https://w.pe./img',
        partyname: 'All peoples party',
        logourl: 'https://app.com/logo'
      }];

      const component = mount(
        <Provider store={store}>
          <Router>
            <Vote candidatesLoaded={true} candidates={candidates} />
          </Router>
        </Provider>
      );
      expect(component).toMatchSnapshot();
      const select = component.find('select');
      expect(select).toHaveLength(1);
      select.simulate('change', { preventDefault() { }})

    });
  });

})
