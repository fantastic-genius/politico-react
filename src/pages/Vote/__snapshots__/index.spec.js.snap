// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Vote /> Rendering should render Vote page correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(Vote) />
</Router>
`;

exports[`<Vote /> Rendering should render Vote page correctly with its children when candidates is loaded 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Vote)
        candidates={
          Array [
            Object {
              "firstname": "john",
              "id": 1,
              "lastname": "mark",
              "logourl": "https://app.com/logo",
              "partyname": "All peoples party",
              "passportUrl": "https://w.pe./img",
            },
          ]
        }
        candidatesLoaded={true}
      >
        <Vote
          candidates={Array []}
          candidatesLoaded={false}
          candidatesLoading={false}
          cleanUp={[Function]}
          error={null}
          getOfficeCandidates={[Function]}
          getOffices={[Function]}
          getVotedCandidates={[Function]}
          offices={Array []}
          token=""
          voteCandidate={[Function]}
          voted={false}
          votedCandidates={Array []}
        >
          <DashboardContainer>
            <div
              className="container dashboard"
            >
              <Connect(DashboardNavBar)>
                <DashboardNavBar
                  signout={[Function]}
                  user={Object {}}
                >
                  <header>
                    <div
                      className="title"
                    >
                      <h3>
                        User Dashboard
                      </h3>
                    </div>
                    <div
                      className="logout-con"
                    >
                      <button
                        className="btn btn-warning"
                        id="logout"
                        onClick={[Function]}
                      >
                        <a
                          href="#"
                        >
                          Logout
                        </a>
                      </button>
                    </div>
                  </header>
                </DashboardNavBar>
              </Connect(DashboardNavBar)>
              <div
                className="main"
              >
                <UserDashboardSideBar>
                  <aside>
                    <Connect(UserInfo)>
                      <UserInfo
                        user={Object {}}
                      >
                        <div
                          className="user"
                        >
                          <div
                            className="photo"
                          >
                            <img
                              alt="User"
                              id="user-img"
                              src="test-file-stub"
                            />
                          </div>
                          <div
                            className="username"
                          >
                            <p
                              id="user-name"
                            >
                              undefined undefined
                            </p>
                          </div>
                        </div>
                      </UserInfo>
                    </Connect(UserInfo)>
                    <div
                      className="menu"
                    >
                      <ul>
                        <li>
                          <Link
                            to="/vote"
                          >
                            <a
                              href="/vote"
                              onClick={[Function]}
                            >
                              Vote
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/myvote"
                          >
                            <a
                              href="/myvote"
                              onClick={[Function]}
                            >
                              My Votes
                            </a>
                          </Link>
                        </li>
                      </ul>
                    </div>
                  </aside>
                </UserDashboardSideBar>
                <article>
                  <div
                    className="title-con"
                  >
                    <h3>
                      Vote Choice Candidate
                    </h3>
                    <hr />
                  </div>
                  <div
                    id="alert"
                  />
                  <div
                    className="select-con"
                  >
                    <label
                      htmlFor="office"
                    >
                      Political Office
                    </label>
                    <select
                      id="office-select"
                      name="office"
                      onChange={[Function]}
                    >
                      <option>
                        Select an Office
                      </option>
                    </select>
                  </div>
                  <div>
                    <table
                      id="candidates-table"
                    >
                      <thead>
                        <tr>
                          <td />
                          <td>
                            Photo
                          </td>
                          <td>
                            Politician Name
                          </td>
                          <td>
                            Party
                          </td>
                          <td>
                            Party Logo
                          </td>
                          <td>
                            Action
                          </td>
                        </tr>
                      </thead>
                      <tbody>
                        <tr />
                      </tbody>
                    </table>
                    <h4>
                      No Candidate for this office
                    </h4>
                  </div>
                </article>
              </div>
              <Footer>
                <footer>
                  <span>
                    Copyright@2019  @fantastic-genius (Andela Bootcamp Cycle41)
                  </span>
                </footer>
              </Footer>
            </div>
          </DashboardContainer>
        </Vote>
      </Connect(Vote)>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`<Vote /> Rendering should render Vote page correctly with its children when candidates is loading 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Vote)
        candidatesLoading={true}
      >
        <Vote
          candidates={Array []}
          candidatesLoaded={false}
          candidatesLoading={false}
          cleanUp={[Function]}
          error={null}
          getOfficeCandidates={[Function]}
          getOffices={[Function]}
          getVotedCandidates={[Function]}
          offices={Array []}
          token=""
          voteCandidate={[Function]}
          voted={false}
          votedCandidates={Array []}
        >
          <DashboardContainer>
            <div
              className="container dashboard"
            >
              <Connect(DashboardNavBar)>
                <DashboardNavBar
                  signout={[Function]}
                  user={Object {}}
                >
                  <header>
                    <div
                      className="title"
                    >
                      <h3>
                        User Dashboard
                      </h3>
                    </div>
                    <div
                      className="logout-con"
                    >
                      <button
                        className="btn btn-warning"
                        id="logout"
                        onClick={[Function]}
                      >
                        <a
                          href="#"
                        >
                          Logout
                        </a>
                      </button>
                    </div>
                  </header>
                </DashboardNavBar>
              </Connect(DashboardNavBar)>
              <div
                className="main"
              >
                <UserDashboardSideBar>
                  <aside>
                    <Connect(UserInfo)>
                      <UserInfo
                        user={Object {}}
                      >
                        <div
                          className="user"
                        >
                          <div
                            className="photo"
                          >
                            <img
                              alt="User"
                              id="user-img"
                              src="test-file-stub"
                            />
                          </div>
                          <div
                            className="username"
                          >
                            <p
                              id="user-name"
                            >
                              undefined undefined
                            </p>
                          </div>
                        </div>
                      </UserInfo>
                    </Connect(UserInfo)>
                    <div
                      className="menu"
                    >
                      <ul>
                        <li>
                          <Link
                            to="/vote"
                          >
                            <a
                              href="/vote"
                              onClick={[Function]}
                            >
                              Vote
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/myvote"
                          >
                            <a
                              href="/myvote"
                              onClick={[Function]}
                            >
                              My Votes
                            </a>
                          </Link>
                        </li>
                      </ul>
                    </div>
                  </aside>
                </UserDashboardSideBar>
                <article>
                  <div
                    className="title-con"
                  >
                    <h3>
                      Vote Choice Candidate
                    </h3>
                    <hr />
                  </div>
                  <div
                    id="alert"
                  />
                  <div
                    className="select-con"
                  >
                    <label
                      htmlFor="office"
                    >
                      Political Office
                    </label>
                    <select
                      id="office-select"
                      name="office"
                      onChange={[Function]}
                    >
                      <option>
                        Select an Office
                      </option>
                    </select>
                  </div>
                  <div>
                    <table
                      id="candidates-table"
                    >
                      <thead>
                        <tr>
                          <td />
                          <td>
                            Photo
                          </td>
                          <td>
                            Politician Name
                          </td>
                          <td>
                            Party
                          </td>
                          <td>
                            Party Logo
                          </td>
                          <td>
                            Action
                          </td>
                        </tr>
                      </thead>
                      <tbody>
                        <tr />
                      </tbody>
                    </table>
                    <h4>
                      No Candidate for this office
                    </h4>
                  </div>
                </article>
              </div>
              <Footer>
                <footer>
                  <span>
                    Copyright@2019  @fantastic-genius (Andela Bootcamp Cycle41)
                  </span>
                </footer>
              </Footer>
            </div>
          </DashboardContainer>
        </Vote>
      </Connect(Vote)>
    </Router>
  </BrowserRouter>
</Provider>
`;
